### PR TITLE
[Feature] Entity Service types

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

Improves and adds types for the Entity Service methods

#### TODO before release

- [x] Merge https://github.com/strapi/strapi/pull/17691
- [x] Merge https://github.com/strapi/strapi/pull/17783

### Why is it needed?

So types are available when using the Entity Service

### How to test it?

Entity Service methods should now be typed

Typescript testing framework has been added in this PR but the tests themselves are still in progress.

### CODE REVIEW

**Review by Convly — DISMISSED** [review]
LGTM

Changes regarding the initial scope
- factories have been reverted to an explicit any type for now to avoid breaking any application while we work around prototype typing issues
- Entities returned by the entity service are strongly typed (properties are narrowed based on given parameters) at the first level, but are keeping every fields for nested structures (relations, etc...)

The last two points will be worked on & shipped in later releases

**Comment by Bassel17** [comment]
we have some failing tests here

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -24,3 +24,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-node@v3
       - uses: ./.github/actions/security/lockfile
+        with:
+          allowedHosts:
```
Comment (Bassel17) [inline .github/workflows/checks.yml:28]: Just a question, am assuming this is to allow ts-zen fork to pass the security check ?

Comment (Convly) [reply .github/workflows/checks.yml:28]: Exactly
btw ping @alexandrebodin

**Review by Convly — APPROVED** [review]

**Review by Convly — APPROVED** [review]

**Comment by jramstedt** [comment]
Good improvements, but createCoreController and createCoreService types are unusable now (and gives typescript errors `Binding element 'strapi' implicitly has an 'any' type.ts(7031)` with strict configs)

**Pull request merged**